### PR TITLE
feat(engine): set the variables locally when failling a job with variables

### LINKED ISSUE #10698: Needs to set the variables locally when failing a job with variables
**Is your feature request related to a problem? Please describe.**

If a failling job has variables, then the variables needs to be set locally

`client.newFailCommand(...).variables(...) `

**Describe the solution you'd like**

Set the variables only locally when failing a job has variables.

**Describe alternatives you've considered**

If you want to variables globally, then you should  `newSetVariablesCommand`, also it can set locally.

- for locally
   `.newSetVariablesCommand(job.getElementInstanceKey()).variables(Map.of("key2", "value2")).local(true).send().join();`
- for globally 
   `.newSetVariablesCommand(job.getElementInstanceKey()).variables(Map.of("key2", "value2").send().join();`

**Additional context**

Related #9404

### PR DESCRIPTION
**Pull request by skayliu**
## Description

If a failing job has variables, then the variables will be set locally.

## Related issues

closes #10698 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by saig0** [comment]
@skayliu thank you for your contribution. :tada: We'll have a look at your PR in the next few days. :eyes:

File: engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobFailProcessor.java [hunk 1]
```diff
@@ -85,6 +91,17 @@ public void afterAccept(
...
+      final DirectBuffer variables = value.getVariablesBuffer();
+      if (variables.capacity() > 0) {
+        variableBehavior.mergeLocalDocument(
+            value.getElementInstanceKey(),
+            value.getProcessDefinitionKey(),
+            value.getProcessInstanceKey(),
+            value.getBpmnProcessIdBuffer(),
+            value.getVariablesBuffer());
+      }
+
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobFailProcessor.java:104]: ❌ This code block in which this code is placed is executed only when the retries is 0 (or negative). The rest of this code block is aimed at creating an incident at the associated task. We should create the variables regardless of the number of retries.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/job/FailJobTest.java [hunk 2]
```diff
@@ -272,4 +273,34 @@ public void shouldRejectFailIfErrorThrown() {
...
+
+    // then
+    Assertions.assertThat(failRecord).hasRecordType(RecordType.EVENT).hasIntent(FAILED);
+    Assertions.assertThat(failRecord.getValue())
+        .hasWorker(job.getWorker())
+        .hasType(job.getType())
+        .hasRetries(retries)
+        .hasDeadline(job.getDeadline());
+
+    assertThat(failRecord.getValue().getVariables()).containsExactly(entry("foo", "bar"));
```
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/job/FailJobTest.java:304]: ❌ This test should also verify that the Variable `foo` is `CREATED` with value `bar` at the variable scope (`failRecord.getValue().getElementInstanceKey()`).

💡 Tip: You can use the `RecordingExporting.variableRecords()` to find the exported variable records.

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks for this @skayliu 🏆 This will be another great contribution!

👍 I really like that you've split up the commits

There are still a few things that need to be addressed, but the main parts are there already. Nice work!

**Review by korthout — APPROVED** [review]
Nice job @skayliu 👏 

I have just one last suggestion. Please let me know if you'd like to apply it. Otherwise, we can just merge this PR.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/job/FailJobTest.java [hunk 3]
```diff
@@ -272,4 +274,45 @@ public void shouldRejectFailIfErrorThrown() {
...
+    final Record<VariableRecordValue> variableRecord =
+        RecordingExporter.variableRecords()
+            .withProcessInstanceKey(job.getProcessInstanceKey())
+            .withScopeKey(failRecord.getValue().getElementInstanceKey())
+            .withName("foo")
+            .getFirst();
+
+    assertThat(variableRecord.getValue().getValue())
+        .describedAs("check set failing job variables locally")
+        .isEqualTo("\"bar\"");
```
[the remark this answers is not in the document]
Comment (skayliu) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/job/FailJobTest.java]: @korthout, Done.

**Comment by korthout** [comment]
bors merge

**Comment by korthout** [comment]
@skayliu We had a problem with our automated merge tooling (bors). It should be fixed after rebasing your changes on top of `main`.

**Pull request merged**